src/IO/ReadBufferFromFileDecorator.cpp

```diff
@@ -0,0 +1,46 @@
+#include <IO/ReadBufferFromFileDecorator.h>
+
+
+namespace DB
+{
+
+ReadBufferFromFileDecorator::ReadBufferFromFileDecorator(std::unique_ptr<ReadBufferFromFileBase> impl_)
+    : impl(std::move(impl_))
+{
+    swap(*impl);
+}
+
+
+std::string ReadBufferFromFileDecorator::getFileName() const
+{
+    return impl->getFileName();
+}
+
+
+off_t ReadBufferFromFileDecorator::getPosition()
+{
+    swap(*impl);
+    auto position = impl->getPosition();
+    swap(*impl);
```

Comment:
I see that this swap was in deleted code, but didn't get why we swap buffers

Reply:
Decorator buffers don't have own buffer in memory, so we need to 'steal' actual working buffer from `impl`, perform an operation and return buffer back.